cdc/log.cc

```diff
@@ -1458,8 +1451,53 @@ class transformer final : public change_processor {
       }
     }
 
-    void set_preimage_select_result(lw_shared_ptr<cql3::untyped_result_set> rs) {
-        _preimage_select_result = std::move(rs);
+    // Note: this assumes that the results are from one partition only
+    void load_preimage_results_into_state(lw_shared_ptr<cql3::untyped_result_set> preimage_set, bool static_only) {
+        // static row
+        if (!preimage_set->empty()) {
+            // There may be some static row data
+            const auto& row = preimage_set->front();
+            for (auto& c : _schema->static_columns()) {
+                if (auto maybe_cell_view = get_preimage_col_value(c, &row)) {
+                    _static_row_state[&c] = *maybe_cell_view;
+                }
+            }
+        }
+
+        if (static_only) {
+            return;
+        }
+
+        // clustering rows
+        for (const auto& row : *preimage_set) {
+            // Construct the clustering key for this row
+            std::vector<bytes> ck_parts;
+            ck_parts.reserve(_schema->clustering_key_size());
+            for (auto& c : _schema->clustering_key_columns()) {
+                auto v = row.get_view_opt(c.name_as_text());
+                if (!v) {
+                    // We might get here if both of the following conditions are true:
+                    // - In preimage query, we requested the static row and some clustering rows,
+                    // - The partition had some static row data, but did not have any requested clustering rows.
+                    // In such case, the result set will have an artificial row that only contains static columns,
+                    // but no clustering columns. In such case, we can safely return from the function,
+                    // as there will be no clustering row data to load into the state.
+                    return;
+                }
+                ck_parts.emplace_back(*v);
+            }
+            auto ck = clustering_key::from_exploded(std::move(ck_parts));
+
+            // Collect regular rows
+            cell_map cells;
+            for (auto& c : _schema->regular_columns()) {
+                if (auto maybe_cell_view = get_preimage_col_value(c, &row)) {
+                    cells[&c] = *maybe_cell_view;
```

Comment:
This concerns me because now we copy the values of cells more than before. This can have a big negative impact on performance.  What's worse it will affect not only batches but also workloads with no batches. Is there a way to optimize this? Maybe we could lazily compute get_col_from_row_state instead of precomputing it.

Reply:
I agree that in current shape this PR adds more copying. I'll try to do something about it.

I'm thinking of writing a consumer object that we could put in place of the one we pass [here](https://github.com/piodul/scylla/blob/cdc-pre-postimage-persistence/cdc/log.cc#L1440). This way we would avoid constructing an `untyped_result_set` object, and instead populate our state directly.

If that doesn't work, I'll try to make initialization lazy.

Reply:
Let's do this as a follow up. This change is complex enough.